xls/passes/rematerialization_pass.cc

```diff
@@ -379,6 +494,68 @@ FindRematerializationOpportunities(
   return result;
 }
 
+// TODO(zihao): replace `FindRematerializationOpportunitiesAtNode` later
+absl::StatusOr<RematEdgeOpportunity>
+FindRematerializationOpportunitiesAtEdge(
+    Node* src, Node* dst, FunctionBase* f,
+    const absl::flat_hash_map<Node*, int64_t>& schedule,
+    const InverseSchedule& inverse_schedule) {
+  int64_t dst_stage_num = schedule.at(dst);
+  absl::flat_hash_set<Node*> available;
+  for (Node* node : inverse_schedule[dst_stage_num]) {
+    available.insert(node);
+  }
+
+  // TODO: remove nodes that would make the schedule worse from `available`
+
+  absl::flat_hash_set<Node*> unavailable;
+  for(int64_t i = 0; i < dst_stage_num; ++i) {
+    for(const auto& node : inverse_schedule[i]){
+      unavailable.insert(node);
+    }
+  }
+
+  absl::flat_hash_set<Node*> chunk = ComputeDeadNodeChunk(unavailable, src);
+  if(chunk.empty() ||
+     !std::any_of(chunk.begin(), chunk.end(),
+                  [](Node* node) { return node->operands().empty(); })){
+    return absl::UnavailableError("Cannot find a proper computation chunk");
+  }
+
+  XLS_ASSIGN_OR_RETURN(auto clones, CloneDeadNodeChunk(f, chunk));
+
+  double quality = RematReducedStageNumber(src, dst, schedule) * 
```

Comment:
Now the evaluated `quality` is more accurate. See the code comments at the top of `RematReducedStageNumber`'s definition.

Reply:
Great stuff, looks correct after a few minutes but I'll need to dig into it heavily to be sure. The other comment about inputs still applies, I think.